x/upgrade/abci.go

```diff
@@ -0,0 +1,40 @@
+package upgrade
+
+import (
+	"fmt"
+
+	sdk "github.com/cosmos/cosmos-sdk/types"
+	abci "github.com/tendermint/tendermint/abci/types"
+)
+
+// BeginBlock will check if there is a scheduled plan and if it is ready to be executed.
+// If it is ready, it will execute it if the handler is installed, and panic/abort otherwise.
+// If the plan is not ready, it will ensure the handler is not registered too early (and abort otherwise).
+//
+// The prupose is to ensure the binary is switch EXACTLY at the desired block, and to allow
+// a migration to be executed if needed upon this switch (migration defined in the new binary)
+func BeginBlock(k *Keeper, ctx sdk.Context, _ abci.RequestBeginBlock) {
+	plan, found := k.GetUpgradePlan(ctx)
+	if !found {
+		return
+	}
+
+	if plan.ShouldExecute(ctx) {
+		if !k.HasHandler(plan.Name) {
+			// We don't have an upgrade handler for this upgrade name, meaning this software is out of date so shutdown
+			ctx.Logger().Error(fmt.Sprintf("UPGRADE \"%s\" NEEDED at %s: %s", plan.Name, plan.DueDate(), plan.Info))
+			panic("UPGRADE REQUIRED!")
+		}
+		// We have an upgrade handler for this upgrade name, so apply the upgrade
+		ctx.Logger().Info(fmt.Sprintf("Applying upgrade \"%s\" at %s", plan.Name, plan.DueDate()))
+		k.ApplyUpgrade(ctx, plan)
```

Comment:
Should disable BlockGasMeter before and after running the apply upgrade since migration changes to stores may exceed block gas limit.

```suggestion
                blockGasMeter := ctx.BlockGasMeter()
                // disable gasmeter for migration
                ctx = ctx.WithBlockGasMeter(sdk.NewInfiniteGasMeter)
		k.ApplyUpgrade(ctx, plan)
		// re-enable gasmeter after migration is done
		ctx = ctx.WithBlockGasMeter(blockGasMeter)
```

Reply:
Good point, I never considered gas in BeginBlock... Assumed it was infinite there and only enforced in the tx context.

Reply:
Also, no need to reset it, as the sub-context is discarded after this function (the input is immutable), but setting it as infinite is good in any case.

Reply:
Yea so there are 2 gas meters in context. The `ctx.GasMeter` is infinite and only set in tx contexts. The `BlockGasMeter` is set with the maximum allowed gas per block parameter before processing the entire block and meters gas for BeginBlock, EndBlock, and tx-processing